qiskit/primitives/containers/observables_array.py

```diff
@@ -71,11 +75,24 @@ def __init__(
         super().__init__()
         if isinstance(observables, ObservablesArray):
             observables = observables._array
+
         self._array = object_array(observables, copy=copy, list_types=(PauliList,))
         self._shape = self._array.shape
+        self._num_qubits = num_qubits
+
         if validate:
             for ndi, obs in np.ndenumerate(self._array):
-                self._array[ndi] = self.coerce_observable(obs)
+                basis_obs = self.coerce_observable(obs)
+                if self._num_qubits is None:
+                    self._num_qubits = basis_obs.num_qubits
+                elif self._num_qubits != basis_obs.num_qubits:
+                    raise ValueError(
+                        "The number of qubits must be the same for all observables in the "
+                        "observables array."
+                    )
+                self._array[ndi] = basis_obs
+        elif self._num_qubits is None and self._array.size > 0:
+            self._num_qubits = self._array.reshape(-1)[0].num_qubits
```

Comment:
Out of scope perhaps, but should I find it weird that the `validate=True` argument doesn't cause `validate()` to be run?

Reply:
The really weird (and undesirable) thing is that an argument named `validate`, a name that sounds innocent and very read-only, can in fact actively modify the object. I don't like it but prefer not to touch it.

I wrote code that called the validate function from the constructor, but it had some bugs that I don't recall, which made be decide to keep not-calling `validate` from the constructor. This means of course that we have here some kind of semi-duplication, where future developers will magically have to be aware that any change in constraints on observable arrays must be dealt with in both the constructor and `validate`.